lib/Transforms/ForwardInsertToExtract/ForwardInsertToExtract.cpp

```diff
@@ -0,0 +1,124 @@
+
+#define DEBUG_TYPE "forward-insert-to-extract"
+
+namespace mlir {
+namespace heir {
+
+#define GEN_PASS_DEF_FORWARDINSERTTOEXTRACT
+#include "lib/Transforms/ForwardInsertToExtract/ForwardInsertToExtract.h.inc"
+
+bool ForwardSingleInsertToExtract::isForwardableOp(
+    Operation *potentialInsert, tensor::ExtractOp &extractOp) const {
+  if (!dominanceInfo.properlyDominates(potentialInsert,
+                                       extractOp.getOperation())) {
+    LLVM_DEBUG(llvm::dbgs() << "insert op does not dominate extract op\n");
+    return false;
+  }
+
+  if (extractOp->getBlock() != potentialInsert->getBlock()) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "insert and extract op are not in the same block\n");
+    return false;
+  }
+
+  return llvm::TypeSwitch<Operation &, bool>(*potentialInsert)
+      .Case<tensor::InsertOp>([&](auto insertOp) {
+        ValueRange insertIndices = insertOp.getIndices();
+        ValueRange extractIndices = extractOp.getIndices();
+        if (insertIndices != extractIndices) {
+          LLVM_DEBUG(llvm::dbgs()
+                     << "insert and extract op do not have matching indices\n");
+          return false;
+        }
+
+        // Naively scan through the operations between the two ops and check if
+        // anything prevents forwarding.
+        for (auto currentNode = insertOp->getNextNode();
+             currentNode != extractOp.getOperation();
+             currentNode = currentNode->getNextNode()) {
+          if (currentNode->getNumRegions() > 0) {
+            LLVM_DEBUG(llvm::dbgs() << "an op with control flow is between the "
+                                       "insert and extract op\n");
+            return false;
+          }
+
+          if (auto op = dyn_cast<tensor::InsertOp>(currentNode)) {
+            if (op.getDest() == insertOp.getDest() &&
+                op.getIndices() == insertIndices) {
+              LLVM_DEBUG(llvm::dbgs()
+                         << "an intermediate op inserts to the same index\n");
+              return false;
+            }
+          }
+        }
+        return true;
+      })
+      .Default([&](Operation &) {
+        LLVM_DEBUG(llvm::dbgs()
+                   << "Unsupported op type, cannot check for forwardability\n");
+        return false;
+      });
+}
+
+FailureOr<Value> getInsertedValue(Operation *insertOp) {
+  return llvm::TypeSwitch<Operation &, FailureOr<Value>>(*insertOp)
+      .Case<tensor::InsertOp>(
+          [&](auto insertOp) { return insertOp.getScalar(); })
+      .Default([&](Operation &) { return failure(); });
+}
+
+LogicalResult ForwardSingleInsertToExtract::matchAndRewrite(
+    tensor::ExtractOp extractOp, PatternRewriter &rewriter) const {
+  LLVM_DEBUG(llvm::dbgs() << "Considering extractOp for replacement: "
+                          << extractOp << "\n");
+
+  for (Operation *use : extractOp.getTensor().getUsers()) {
```

Comment:
Thank you so much for starting this pass! 

I actually think the only issue is from this line. I ran this pass with `--debug` on the testcase you added and noticed that the extract operation that could be replaced wasn't picking up the forwardable insert operation:
```
  * Pattern mlir::heir::ForwardSingleInsertToExtract : 'tensor.extract -> ()' {
Trying to match "mlir::heir::ForwardSingleInsertToExtract"
Considering extractOp for replacement: %extracted_2 = tensor.extract %inserted[%c0, %c0] : tensor<1x16x!lwe.rlwe_ciphertext<encoding = #lwe.polynomial_evaluation_encoding<cleartext_start = 32, cleartext_bitwidth = 32>, rlwe_params = <ring = <coefficientType = i32, coefficientModulus = 463187969 : i32, polynomialModulus = <1 + x**16>>>, underlying_type = f32>>
Considering use for forwarding: %inserted_3 = tensor.insert %5 into %inserted[%c0, %c0] : tensor<1x16x!lwe.rlwe_ciphertext<encoding = #lwe.polynomial_evaluation_encoding<cleartext_start = 32, cleartext_bitwidth = 32>, rlwe_params = <ring = <coefficientType = i32, coefficientModulus = 463187969 : i32, polynomialModulus = <1 + x**16>>>, underlying_type = f32>>
insert op does not dominate extract op
Use is not forwardable: %inserted_3 = tensor.insert %5 into %inserted[%c0, %c0] : tensor<1x16x!lwe.rlwe_ciphertext<encoding = #lwe.polynomial_evaluation_encoding<cleartext_start = 32, cleartext_bitwidth = 32>, rlwe_params = <ring = <coefficientType = i32, coefficientModulus = 463187969 : i32, polynomialModulus = <1 + x**16>>>, underlying_type = f32>>
Considering use for forwarding: %extracted_2 = tensor.extract %inserted[%c0, %c0] : tensor<1x16x!lwe.rlwe_ciphertext<encoding = #lwe.polynomial_evaluation_encoding<cleartext_start = 32, cleartext_bitwidth = 32>, rlwe_params = <ring = <coefficientType = i32, coefficientModulus = 463187969 : i32, polynomialModulus = <1 + x**16>>>, underlying_type = f32>>
insert op does not dominate extract op
Use is not forwardable: %extracted_2 = tensor.extract %inserted[%c0, %c0] : tensor<1x16x!lwe.rlwe_ciphertext<encoding = #lwe.polynomial_evaluation_encoding<cleartext_start = 32, cleartext_bitwidth = 32>, rlwe_params = <ring = <coefficientType = i32, coefficientModulus = 463187969 : i32, polynomialModulus = <1 + x**16>>>, underlying_type = f32>>
"mlir::heir::ForwardSingleInsertToExtract" result 0
```

This is because a `tensor.insert` operation (`%inserted = tensor.insert %2 into %arg2[%c0, %c0] : tensor<1x16x!lwe.rlwe_ciphertext<>>
` produces a *new* tensor `%inserted`, which means that the only operations picked up by this statement are the operations that use the newly inserted tensor. So I think this logic could be very simple - you might just want to check if `extractOp.getTensor().getDefiningOp<tensor::InsertOp>` exists and then replace with the scalar.

Reply:
These **good first issue**s are fun :blush:

Reply:
I know! Usually they sound easy and then there's something sneaky in there that I didn't anticipate either :P But at least we all learn